### PR TITLE
Rewarded ads

### PR DESCRIPTION
**Pull request by JeffreyQ**
Added backend support for rewarded (video) ads

### CODE REVIEW

File: app/controllers/impressions_controller.rb [hunk 1]
```diff
@@ -39,6 +39,14 @@ def update
     end
   end
 
+  def ratio_interacted
+    @impression = Impression.find(params[:id])
+    @ad_unit = AdUnit.find(params[:id])
```
Comment (JeffreyQ) [inline app/controllers/impressions_controller.rb]: "placeholder" logic for time being, until we get impressions controller up and running

File: app/controllers/impressions_controller.rb [hunk 2]
```diff
@@ -39,6 +39,14 @@ def update
     end
   end
 
+  def ratio_interacted
+    @impression = Impression.find(params[:id])
```
Comment (johnli611) [inline app/controllers/impressions_controller.rb]: reference @impression directly thanks to rails magic

File: app/controllers/impressions_controller.rb (hunk 1 above)
Comment (johnli611) [inline app/controllers/impressions_controller.rb]: @impression.ad_unit

File: app/controllers/impressions_controller.rb [hunk 3]
```diff
@@ -39,6 +39,14 @@ def update
...
+    @ad_unit = AdUnit.find(params[:id])
+    if @ad_unit.rewarded
```
Comment (johnli611) [inline app/controllers/impressions_controller.rb]: @impression.ad_unit.rewarded

File: app/controllers/impressions_controller.rb (hunk 1 above)
Comment (johnli611) [reply app/controllers/impressions_controller.rb]: delete line

File: app/controllers/impressions_controller.rb (hunk 2 above)
Comment (johnli611) [reply app/controllers/impressions_controller.rb]: delete

File: app/controllers/impressions_controller.rb [hunk 4]
```diff
@@ -39,6 +39,14 @@ def update
...
+      return @impression.interaction_length / @ad_unit.video_length
```
Comment (johnli611) [inline app/controllers/impressions_controller.rb]: Will this return an int always? Double check

File: app/models/ad_unit.rb [hunk 5]
```diff
@@ -27,12 +29,22 @@ class AdUnit < ApplicationRecord
...
+  scope :interstitial, -> { where(interstitial: true) }
+  scope :noninterstitial, -> { where(interstitial: false) }
+
 
   ### VALIDATIONS ###
 
   validates :title, :ad_unit_url, :dimensions, presence: true
   validates :ad_format, inclusion: VALID_FORMATS, presence: true
   validates :dimensions, inclusion: VALID_DIMENSIONS, presence: true
+  validates :ad_format, inclusion: [FORMAT_VIDEO], presence: true, if: :ad_is_rewarded?
```
Comment (johnli611) [inline app/models/ad_unit.rb]: see if you can make this just `if: :rewarded`

File: db/migrate/20180522224910_add_rewarded_ads_columns.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class AddRewardedAdsColumns < ActiveRecord::Migration[5.1]
+  def change
+    add_column :ad_units, :rewarded, :boolean
+    add_column :ad_units, :interstitial, :boolean
+    # video_length measured in milliseconds
```
Comment (johnli611) [inline db/migrate/20180522224910_add_rewarded_ads_columns.rb]: delete comment

File: app/controllers/impressions_controller.rb [hunk 7]
```diff
@@ -39,6 +39,14 @@ def update
     end
   end
 
+  def ratio_interacted
```
Comment (johnli611) [inline app/controllers/impressions_controller.rb]: should move to model

File: spec/factories/ad_unit.rb [hunk 8]
```diff
@@ -9,5 +9,8 @@
     ad_format { AdUnit::FORMAT_VIDEO }
     last_served_at nil
     active true
+    rewarded false
+    interstitial false
+    video_length 0
```
Comment (johnli611) [inline spec/factories/ad_unit.rb]: add trait for `:video`, `:interstitial`, `:rewarded`
`https://robots.thoughtbot.com/remove-duplication-with-factorygirls-traits`

File: app/controllers/impressions_controller.rb (hunk 7 above)
Comment (johnli611) [reply app/controllers/impressions_controller.rb]: write a test

File: spec/factories/impression.rb [hunk 9]
```diff
@@ -6,6 +6,7 @@
     clicked false
     ad_unit
     app_session
+    interaction_length 0
 
     trait :shown do
```
Comment (JeffreyQ) [inline spec/factories/impression.rb]: `testing`

File: app/models/impression.rb [hunk 10]
```diff
@@ -9,6 +9,18 @@ class Impression < ApplicationRecord
...
   ### VALIDATIONS ###
 
+  def ratio_interacted
+    if self.interaction_length == nil || self.ad_unit.video_length == nil || self.ad_unit.video_length <= 0
+      return nil
+    end
+
+    if self.ad_unit.rewarded
+      return self.interaction_length.to_f / self.ad_unit.video_length.to_f
+    else
```
Comment (JeffreyQ) [inline app/models/impression.rb]: remove else

File: app/models/impression.rb [hunk 11]
```diff
@@ -9,6 +9,18 @@ class Impression < ApplicationRecord
 
   ### VALIDATIONS ###
 
+  def ratio_interacted
+    if self.interaction_length == nil || self.ad_unit.video_length == nil || self.ad_unit.video_length <= 0
+      return nil
+    end
+
+    if self.ad_unit.rewarded
+      return self.interaction_length.to_f / self.ad_unit.video_length.to_f
```
Comment (JeffreyQ) [inline app/models/impression.rb]: remove return

File: app/models/impression.rb [hunk 12]
```diff
@@ -9,6 +9,18 @@ class Impression < ApplicationRecord
 
   ### VALIDATIONS ###
 
+  def ratio_interacted
+    if self.interaction_length == nil || self.ad_unit.video_length == nil || self.ad_unit.video_length <= 0
```
Comment (JeffreyQ) [inline app/models/impression.rb]: reduce to single line

File: spec/factories/ad_unit.rb [hunk 13]
```diff
@@ -9,5 +9,15 @@
     ad_format { AdUnit::FORMAT_VIDEO }
     last_served_at nil
     active true
+    video_length 0
```
Comment (JeffreyQ) [inline spec/factories/ad_unit.rb]: change video_length

File: spec/factories/ad_unit.rb [hunk 14]
```diff
@@ -9,5 +9,15 @@
...
+
+    trait :rewarded do
+      rewarded true
+    end
+
+    trait :interstitial do
+      interstitial true
+    end
+
```
Comment (JeffreyQ) [inline spec/factories/ad_unit.rb]: remove extra space

File: spec/factories/impression.rb [hunk 15]
```diff
@@ -20,5 +21,6 @@
       clicked true
       clicked_at Time.now
     end
+
```
Comment (JeffreyQ) [inline spec/factories/impression.rb]: remove extra space

File: spec/models/impression_spec.rb [hunk 16]
```diff
@@ -12,4 +12,26 @@
...
+      impression.ad_unit.update(video_length: 30000)
+      impression.ad_unit.update(rewarded: true)
+
+      expect(impression.ratio_interacted).to eql 0.5
+    end
+
+    it "should not return a ratio without a video_length value" do
+      impression.ad_unit.update(rewarded: true)
+
+      expect(impression.ratio_interacted).to eql nil
```
Comment (JeffreyQ) [inline spec/models/impression_spec.rb]: change to be_nil

File: spec/models/impression_spec.rb [hunk 17]
```diff
@@ -12,4 +12,26 @@
...
+    end
+
+    it "should not return a ratio if ad_unit is not rewarded" do
+      expect(impression.ratio_interacted).to eql nil
```
Comment (JeffreyQ) [inline spec/models/impression_spec.rb]: change to be_nil

**Pull request merged**